File: adobe-dev/leetcode-top-75/questions/linked-list/linked-list/remove-nth-from-end/solution.py
import os
import sys
from typing import List, Dict, Optional, Tuple

ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../"))
if ROOT_DIR not in sys.path:
    sys.path.append(ROOT_DIR)

from shared.python.ds import ArrayList, ListNode, Stack, Queue, TreeNode, MinHeap, MaxHeap

def remove_nth_from_end(head: Optional[ListNode], n: int) -> Optional[ListNode]:
    dummy = ListNode(0, head)
    fast = dummy
    slow = dummy
    i = 0
    while i < n + 1:
        fast = fast.next
        i += 1
    while fast:
        fast = fast.next
        slow = slow.next
    slow.next = slow.next.next
    return dummy.next

File: adobe-dev/leetcode-top-75/questions/linked-list/linked-list/remove-nth-from-end/solution.ts
import {
  ArrayList,
  Stack,
  Queue,
  ListNode,
  TreeNode,
  MinHeap,
  MaxHeap,
} from "../../../../shared/typescript/ds";

export const removeNthFromEnd = (head: ListNode<number> | null, n: number): ListNode<number> | null => {
  const dummy = new ListNode<number>(0, head);
  let fast: ListNode<number> | null = dummy;
  let slow: ListNode<number> | null = dummy;
  for (let i = 0; i < n + 1; i += 1) {
    fast = fast?.next ?? null;
  }
  while (fast !== null) {
    fast = fast.next;
    slow = slow?.next ?? null;
  }
  if (slow !== null && slow.next !== null) slow.next = slow.next.next;
  return dummy.next;
};
